core/src/main/java/org/elasticsearch/discovery/zen/publish/PublishClusterStateAction.java

```diff
@@ -245,63 +333,290 @@ public static BytesReference serializeDiffClusterState(Diff diff, Version nodeVe
+        }
+
+        try {
+            listener.onNewClusterState(committedClusterState, new NewClusterStateListener.NewStateProcessed() {
+                @Override
+                public void onNewClusterStateProcessed() {
+                    try {
+                        channel.sendResponse(TransportResponse.Empty.INSTANCE);
+                    } catch (Throwable e) {
+                        logger.debug("failed to send response on cluster state processed", e);
+                        onNewClusterStateFailed(e);
+                    }
                 }
-                lastSeenClusterState.status(ClusterState.ClusterStateStatus.RECEIVED);
+
+                @Override
+                public void onNewClusterStateFailed(Throwable t) {
+                    try {
+                        channel.sendResponse(t);
+                    } catch (Throwable e) {
+                        logger.debug("failed to send response on cluster state processed", e);
+                    }
+                }
+            });
+        } catch (Exception e) {
+            logger.warn("unexpected error while processing cluster state version [{}]", e, lastSeenClusterState.version());
+            throw e;
+        }
+    }
+
+    private class SendClusterStateRequestHandler implements TransportRequestHandler<BytesTransportRequest> {
+
+        @Override
+        public void messageReceived(BytesTransportRequest request, final TransportChannel channel) throws Exception {
+            handleIncomingClusterStateRequest(request, channel);
+        }
+    }
+
+    private class CommitClusterStateRequestHandler implements TransportRequestHandler<CommitClusterStateRequest> {
+        @Override
+        public void messageReceived(CommitClusterStateRequest request, final TransportChannel channel) throws Exception {
+            handleCommitRequest(request, channel);
+        }
+    }
+
+    protected static class CommitClusterStateRequest extends TransportRequest {
+
+        String stateUUID;
+
+        public CommitClusterStateRequest() {
+        }
+
+        public CommitClusterStateRequest(String stateUUID) {
+            this.stateUUID = stateUUID;
+        }
+
+        @Override
+        public void readFrom(StreamInput in) throws IOException {
+            super.readFrom(in);
+            stateUUID = in.readString();
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            super.writeTo(out);
+            out.writeString(stateUUID);
+        }
+    }
+
+
+    class SendingController {
+
+        private final ClusterState clusterState;
+
+        public BlockingClusterStatePublishResponseHandler getPublishResponseHandler() {
+            return publishResponseHandler;
+        }
+
+        private final BlockingClusterStatePublishResponseHandler publishResponseHandler;
```

Comment:
I am not sure it is a good idea to make this a member of sendingController. It is mostly used in `PublishClusterStateAction` anyway and only when a node fails to respond to the cluster state sending we use it in SendingController. This makes it slightly more difficult to read the code because now I have to look in two places to see how this handler is actually used. Can we remove it here and just use it in PublishClusterStateAction?